### ISSUE TITLE
clang -O0 miscompilation on aarch64-darwin

### ISSUE THREAD

**Issue by TerrorJack** [llvm/llvm-project#89112]
Reproducer:

```c
#include <stdio.h>
#include <stdint.h>
#include <stdatomic.h>
#include <stdbool.h>

int main (void) {
    _Atomic uint32_t x = 42;
    uint32_t y = 42;

    bool success = atomic_compare_exchange_strong(&x, &y, 43);
    printf("# CAS\n");
    printf("success=%u\n", (int) success);
    printf("old=%u\n", y);
    printf("x=%u\n", x);

    printf("# Swap\n");
    y = atomic_exchange(&x, 2);
    printf("x=%u\n", x);
    printf("y=%u\n", y);

    printf("# Fetch-Add\n");
    y = atomic_fetch_add(&x, 2);
    printf("x=%u\n", x);
    printf("y=%u\n", y);
    fflush(stdout);
}
```

The correct runtime output should be:

```
# CAS
success=1
old=42
x=43
# Swap
x=2
y=43
# Fetch-Add
x=4
y=2
```

On a macos-14 m1 machine, using `cc` from apple sdk compiles the given reproducer fine. However, when using `clang` compiled from `llvmorg-18.1.4`, the runtime output of `clang -O0 test.c -o test` is:

```
# CAS
success=1
old=42
x=536870912
# Swap
x=536870912
y=43
# Fetch-Add
x=536870912
y=2
```

The runtime result is correct with `-O1` and above, but something is still definitely wrong! Problematic `clang` version:

```
clang version 18.1.4 (git@github.com:llvm/llvm-project.git e6c3289804a67ea0bb6a86fadbe454dd93b8d855)
Target: arm64-apple-darwin23.4.0
Thread model: posix
InstalledDir: /Users/terrorjack/workspace/llvm-project/build/bin
```

**Comment by shafik** [comment]
I am not able to reproduce on godbolt: https://godbolt.org/z/xeGK3nWc3 or locally on my M1.  

CC @Bigcheese @rjmccall

**Comment by TerrorJack** [comment]
I tried to bisect it and it seems the first commit that introduces the miscompilation is ff47989ec238dafe4a68c6a716e8dbccc9f559f5. The exact command I used to configure the LLVM tree is:

```
cmake \
  -B build \
  -DCMAKE_BUILD_TYPE=Release \
  -DDEFAULT_SYSROOT=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk \
  -DLLVM_CCACHE_BUILD=ON \
  -DLLVM_ENABLE_LIBXML2=OFF \
  -DLLVM_ENABLE_PIC=OFF \
  -DLLVM_ENABLE_PROJECTS="clang" \
  -DLLVM_ENABLE_RUNTIMES="compiler-rt" \
  -DLLVM_ENABLE_TERMINFO=OFF \
	-DLLVM_ENABLE_ZLIB=OFF \
	-DLLVM_ENABLE_ZSTD=OFF \
  -DLLVM_PARALLEL_LINK_JOBS=1 \
  -DCMAKE_EXPORT_COMPILE_COMMANDS=ON \
  -G Ninja \
  llvm
```

**Comment by shafik** [comment]
CC @aemerson

**Comment by aemerson** [comment]
Sorry, I missed this. I'll take a look.

**Comment by aemerson** [comment]
Actually this doesn't repro on trunk anymore, can you verify?

**Comment by TerrorJack** [comment]
Ah indeed! I ran a bisect again and seems it's fixed since 60fc4ac67a613e4e36cef019fb2d13d70a06cfe8. @aemerson Any chance it may be backported to 18.x?

**Comment by aemerson** [comment]
@tstellar could we backport this? I'm not sure what the right command is to make it happen.
